applications/nrf5340_audio/src/modules/display/tab_log.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/display.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/zbus/zbus.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_ctrl.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+
+#include <lvgl.h>
+#include "tab_log.h"
+
+#define LOG_MSG_MAX_LENGTH	       200
+#define LOG_MSG_N_TIMESTAMP_MAX_LENGTH 220
+
+LOG_MODULE_DECLARE(display_log, CONFIG_DISPLAY_LOG_LEVEL);
+
+K_MUTEX_DEFINE(nrf5340_display_mtx); /* This is used so there are no conflicts between scheduler and
+					LVGL */
+static lv_obj_t *tab_log;
+int log_list_length;
+
+bool stop_logging;
+struct backend_context {
+	uint32_t cnt;
+	const char *exp_str[10];
+	uint32_t delay_ms;
+	bool active;
+	struct k_timer timer;
+};
+struct backend_context display_log_context;
+
+/**
+ * @brief  putc implementation for logging backend process
+ *
+ * @details See documentation for cbprint_cb function pointer in zephyr\sys\cbprintf.h
+ */
+static int cbprintf_callback(int c, void *ctx)
+{
+	char **p = ctx;
+
+	**p = c;
+	(*p)++;
+	return c;
+}
+
+static void log_panic_handler(const struct log_backend *const backend)
+{
+	ARG_UNUSED(backend);
+}
+
+static void expire_cb(struct k_timer *timer)
+{
+	void *ctx = k_timer_user_data_get(timer);
+	struct backend_context *context = (struct backend_context *)ctx;
+
+	context->active = true;
+}
+
+static void backend_init(const struct log_backend *const backend)
+{
+	struct backend_context *context = (struct backend_context *)backend->cb->ctx;
+
+	k_timer_init(&context->timer, expire_cb, NULL);
+	k_timer_user_data_set(&context->timer, (void *)context);
+	k_timer_start(&context->timer, K_MSEC(context->delay_ms), K_NO_WAIT);
+}
+
+static int backend_is_ready(const struct log_backend *const backend)
+{
+	struct backend_context *context = (struct backend_context *)backend->cb->ctx;
+
+	return context->active ? 0 : -EBUSY;
+}
+
+static void add_log_msg_to_log_list(lv_obj_t *tab_log, char *log_msg, lv_color_t log_msg_color)
+{
+	lv_obj_t *label = lv_label_create(tab_log);
+
+	lv_label_set_long_mode(label, LV_LABEL_LONG_WRAP);
+	lv_obj_set_width(label, 290);
+	lv_label_set_text(label, log_msg);
+	lv_obj_set_style_text_color(label, log_msg_color, 0);
+	lv_obj_scroll_to_view_recursive(label, LV_ANIM_OFF);
+}
+
+static void backend_process(const struct log_backend *const backend, union log_msg_generic *msg)
+{
+	log_timestamp_t msg_timestamp;
+	int msg_timestamp_in_us;
+	char str[LOG_MSG_MAX_LENGTH];
+	char timestamp_n_str[LOG_MSG_N_TIMESTAMP_MAX_LENGTH];
+	char *pstr = str;
+
+	lv_obj_t *oldest_label;
```

Comment:
This can actually be declared on line 161 inside the if clause where it's used since that is the only scope it's used within